// multiformats/dist/src/bytes.js
export const empty = new Uint8Array(0);
export function toHex(d) {
    return d.reduce((hex, byte) => hex + byte.toString(16).padStart(2, '0'), '');
}
export function fromHex(hex) {
    const hexes = hex.match(/../g);
    return hexes != null ? new Uint8Array(hexes.map(b => parseInt(b, 16))) : empty;
}
export function equals(aa, bb) {
    if (aa === bb) {
        return true;
    }
    if (aa.byteLength !== bb.byteLength) {
        return false;
    }
    for (let ii = 0; ii < aa.byteLength; ii++) {
        if (aa[ii] !== bb[ii]) {
            return false;
        }
    }
    return true;
}
/**
 * Normalize binary input to a plain `Uint8Array` backed by an `ArrayBuffer`.
 *
 * Returns the input itself when it is already a plain `Uint8Array` over an
 * `ArrayBuffer`, otherwise a fresh view (or, for `SharedArrayBuffer`-backed
 * input, a copy) over the same bytes.
 *
 * Throws if input is not a recognised binary type.
 */
export function coerce(o) {
    if (o instanceof Uint8Array && o.constructor.name === 'Uint8Array') {
        return toArrayBufferBackedArray(o);
    }
    if (o instanceof ArrayBuffer) {
        return new Uint8Array(o);
    }
    if (ArrayBuffer.isView(o)) {
        return toArrayBufferBackedArray(new Uint8Array(o.buffer, o.byteOffset, o.byteLength));
    }
    throw new Error('Unknown type, must be binary type');
}
export function isBinary(o) {
    return o instanceof ArrayBuffer || ArrayBuffer.isView(o);
}
export function fromString(str) {
    return new TextEncoder().encode(str);
}
export function toString(b) {
    return new TextDecoder().decode(b);
}
function isByteArrayWithArrayBuffer(b) {
    return b?.buffer instanceof ArrayBuffer;
}
/**
 * Ensures `b` is backed by an ArrayBuffer - if not a new Uint8Array will be
 * created and the contents of `b` copied into it.
 */
export function toArrayBufferBackedArray(b) {
    if (isByteArrayWithArrayBuffer(b)) {
        return b;
    }
    return b.slice();
}
